### PR TITLE
[Console] Add support for JSON with long numerals

### LINKED ISSUE #4536: [Decouple] [BUG] Dev tool console is losing big number precision due to client
**Describe the bug**

Dev tool console is losing precision when displaying big number.

**To Reproduce**
1. Go to dev tool console in OSD
2. Index one `long` type data into OS, and send GET request to retrieve it.
    ```
    PUT index_0001
    {
    "mappings": {
      "properties": {
        "test_long": {
          "type": "long"
        }
      }
    }
    }
    
    PUT index_0001/_doc/0
    {
    "test_long": 1688275058533285124
    }
    
    GET index_0001/_search
    ```
4. You'll get below results
    ```
    {
      "took": 58,
      "timed_out": false,
      "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
      },
      "hits": {
        "total": {
          "value": 1,
          "relation": "eq"
        },
        "max_score": 1,
        "hits": [
          {
            "_index": "index_0001",
            "_id": "0",
            "_score": 1,
            "_source": {
              "test_long": 1688275058533285000
            }
          }
        ]
      }
    }
    ```
6. We can see it loses precision comparing original data `1688275058533285124` and the displayed data in OSD `1688275058533285000`
7. Also I validated the actual data in OS is stored with correct precision.

**Expected behavior**
OSD should display the same number as stored in OS, without losing precision

**OpenSearch Version**
2.5 and later

**Dashboards Version**
2.5 and later

**Additional context**

JavaScript does not support 64 bit integer values. Using a number in JavaScript results in the Number type which itself is a 64bit IEEE floating point number, thus leaving us a 52 bit mantisse to store integer values without rounding errors. Everything above what fits into 52 bit will have rounding errors.

Thus fields in OpenSearch that can contain numbers outside that safe JavaScript range(E.g. `long`)  is prone to that rounding error in Dev tool console

### PR DESCRIPTION
**Pull request by AMoo-Miki**
Add support for JSON with long numerals in Dev Tools.

Also:
* Add support for parsing and stringifying JSON with long numerals into `@osd/std`
* Upgrade `@opensearch/opensearch@2.3.0` which supports long numerals
* Add support for long numerals to `http/fetch`

### Description

* When long numerals were used in the editor, they lost precision when 
  * requests were run
  * text was auto-indented
* When a result contained long numerals, it lost precision when displayed in console.

### Issues Resolved

Fixes #4536 

### Before

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/3527403/3a72c3c7-9928-471e-b6d0-d0a20762d61b

### After

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/3527403/6bb45a36-a8d4-42fe-98b8-6a6f4fc01e55

### Check List

- [X] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by ashwin-pc** [comment]
Nice addition, can you add documentation for the change in the package @AMoo-Miki ?

**Comment by zhongnansu** [comment]
re-running failed CIs

**Comment by zhongnansu** [comment]
@AMoo-Miki Hi Miki, Would you check the CI failures? Is it because we haven't release the js client with the fix?

**Comment by AMoo-Miki** [comment]
Thanks. This needs to be bumped to use the 2.3.1 client.

**Comment by zhongnansu** [comment]
> 

js-client 2.3.1 is out. Would you update the PR? Thx! @AMoo-Miki

**Comment by zhongnansu** [comment]
> Thanks. This needs to be bumped to use the 2.3.1 client.


js client 2.3.1 is out. Would you update the PR? Thx! @AMoo-Miki

**Comment by AMoo-Miki** [comment]
I am stuck with making a decision for how this should be implemented as we don't want to add it all over OSD. Will continue thinking.

**Comment by ashwin-pc** [comment]
@AMoo-Miki should i remove the 2.10 tag for this then?

**Comment by AMoo-Miki** [comment]
> @AMoo-Miki should i remove the 2.10 tag for this then?

Nah. Just needed some time to work on it :)

**Review by kavilla — APPROVED** [review]
The backport will have to be custom correct?

**Comment by AMoo-Miki** [comment]
> The backport will have to be custom correct?

I am hoping it won't 🤞

File: packages/osd-std/README.md [hunk 1]
```diff
@@ -1,3 +1,73 @@
 # `@osd/std` — OpenSearch Dashboards standard library
 
-This package is a set of utilities that can be used both on server-side and client-side.
\ No newline at end of file
+This package is a set of utilities that can be used both on server-side and client-side.
+
+## API
```
Comment (joshuarrrr) [inline packages/osd-std/README.md:5]: thanks for adding!

File: packages/osd-std/src/json.ts [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+ *    8) 9000 - 9999: [9-9]\d{3}
+ *    Finally we add anything 5 digits or longer: `\d{5,}
+ *
+ * Note: A better solution would use AST but considering its performance penalty, RegExp is the next
+ * best thing.
+ */
+const maxIntAsString = String(Number.MAX_SAFE_INTEGER);
+const maxIntLength = maxIntAsString.length;
+// Sub-patterns for each digit
+const longNumeralMatcherTokens = [`\\d{${maxIntAsString.length + 1},}`];
```
Comment (joshuarrrr) [inline packages/osd-std/src/json.ts:37]: nit
```suggestion
const longNumeralMatcherTokens = [`\\d{${maxIntLength + 1},}`];
```

File: packages/osd-std/src/json.ts [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+  const { marker } = getMarker(candidate);
+
+  /* The matcher that looks for "<marker><numerals>"
+   * Because we have made sure that `marker` was never present in the original object, we can
+   * carelessly assume every "<marker><numerals>" is due to our marking.
+   */
+  const markedBigIntMatcher = new RegExp(`"${marker}(-?\\d+)"`, 'g');
+
+  /* Convert BigInt values to a string and mark them.
+   * Can't be bothered with Number values outside the safe range because they are already corrupted.
```
Comment (joshuarrrr) [inline packages/osd-std/src/json.ts:212]: Do we want to throw some sort of error in that case?

**Review by joshuarrrr — APPROVED** [review]

File: packages/osd-std/src/json.ts (hunk 2 above)
Comment (AMoo-Miki) [reply packages/osd-std/src/json.ts:37]: Thanks. Will update in a separate PR.

File: packages/osd-std/src/json.ts (hunk 3 above)
Comment (AMoo-Miki) [reply packages/osd-std/src/json.ts:212]: Those long ones would have already lost precision so letting them pass through will not ruin anything.

**Review by ashwin-pc — COMMENTED** [review]
Just one concern about how we are exposing the custom client. The rest are just minor nits or opinions, feel free to use them if you think it makes sense :)

File: packages/osd-std/src/json.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply packages/osd-std/src/json.ts:37]: I explained why not in the comments in the code. Before parsing the long string, I wouldn't know what's in it. Assuming 8192 is max int, if i take everything greater than 8000 and turn them into BigInt, that would be unexpected to everyone who don't use numbers bigger than 8192. Hence, doing this one time activity of making a marginally more complex regex.

File: packages/osd-std/src/json.ts [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+  /* Convert marked values to BigInt values.
+   * The `startsWith` is purely for performance, to avoid running `test` if not needed.
+   */
+  const convertMarkedValues = (val: any) =>
+    typeof val === 'string' && val.startsWith(marker) && markedValueMatcher.test(val)
+      ? BigInt(val.substring(length))
+      : val;
+
+  /* For better performance, instead of testing for existence of `reviver` on each value, two almost
```
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply packages/osd-std/src/json.ts:130]: The code could be
```
loop {
  condition ? doA : doB
}
```

but then condition is evaluated repeatedly. I decided to do
```
condition ? loop { doA } : loop { doB }
```
this results in a little less maintainable code.

File: packages/osd-std/src/json.ts [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    }
+  })(0, 0);
+
+  return choices;
+};
+
+/* Experiments with different combinations of various lengths, until one is found to not be in
+ * the input string.
+ */
+const getMarker = (text: string): { marker: string; length: number } => {
```
[the remark this answers is not in the document]
Comment (AMoo-Miki) [reply packages/osd-std/src/json.ts:92]: What if a value uses UUIDs? also, the longer the UUID, the more expensive my regex will be. Some more details [here](https://github.com/opensearch-project/opensearch-js/pull/544).

**Comment by kavilla** [comment]
@AMoo-Miki I have reverted this on 2.10 and tagged with 2.11. The reason being I'm not positive if the JS client cherry pick commit https://github.com/opensearch-project/opensearch-js/pull/550 is intended to make the client 2.x version to be non-breaking for major version bump. So that we can utilize the declaration of `opensearch` instead of using `opensearch-next`

**Comment by kavilla** [comment]
@AMoo-Miki will address the backport

**Pull request merged**